core/src/controls.ts

```diff
@@ -16,82 +16,118 @@
+ * @returns A new `ControlSignal` object based on `spec`.
+ */
+export function controlSignalFromSpec(spec: BinaryCounterSpec): BinaryCounter;
+export function controlSignalFromSpec(spec: ChordProgressionSpec):
+    ChordProgression;
+export function controlSignalFromSpec(spec: ControlSignalSpec) {
+  switch (spec.type) {
+    case 'BinaryCounter':
+      return new BinaryCounter(spec.args);
+    case 'ChordProgression':
+      return new ChordProgression(spec.args);
+    default:
+      throw new Error(`Unknown control signal: ${spec}`);
+  }
+}
 
 /**
  * Abstract ControlSignal class for producing control tensors.
  *
- * @param numSteps The length of the control signal.
  * @param depth The size of the control tensor at each step.
  */
-export abstract class ControlSignal {
-  readonly numSteps: number;  // Total length of sequences.
-  readonly depth: number;     // Size of final output dimension.
-  abstract getTensor(step: number): tf.Tensor1D;
+export abstract class ControlSignal<T> {
+  readonly depth: number;  // Size of final output dimension.
+  abstract getTensors(numSteps: number, args: T): tf.Tensor2D;
 
-  constructor(numSteps: number, depth: number) {
-    this.numSteps = numSteps;
-    this.depth = depth;
-  }
+  constructor(depth: number) { this.depth = depth; }
 }
 
 /**
  * Binary counter control signal.
  *
- * @param numSteps The length of the control signal.
  * @param numBits The number of binary counter bits.
  */
-export class BinaryCounter extends ControlSignal {
-  constructor(numSteps: number, numBits: number) { super(numSteps, numBits); }
+export class BinaryCounter extends ControlSignal<BinaryCounterUserArgs> {
+  constructor(args: BinaryCounterArgs) { super(args.numBits); }
 
   /**
-   * Get the tensor of control values at the specified step. The returned tensor
-   * acts as a binary counter (using -1 instead of 0), with least significant
-   * bit first.
+   * Get the tensor of control values. The returned tensor acts as a binary
+   * counter (using -1 instead of 0), with least significant bit first.
    *
-   * @param step Step for which to get the binary counter.
+   * @param numSteps Number of steps of binary counter values to return.
+   * @returns A 2D tensor of binary counter values.
    */
-  getTensor(step: number) {
-    if (step >= this.numSteps) {
-      throw new RangeError(`Step out of range: ${step} >= ${this.numSteps}`);
+  getTensors(numSteps: number, args: BinaryCounterUserArgs) {
+    const buffer = tf.buffer([numSteps, this.depth]);
+    for (let step = 0; step < numSteps; ++step) {
+      for (let i = 0; i < this.depth; ++i) {
+        buffer.set(Math.floor(step / Math.pow(2, i)) % 2 ? 1.0 : -1.0, step, i);
+      }
     }
-
-    const buffer = tf.buffer([this.depth]);
-    for (let i = 0; i < this.depth; ++i) {
-      buffer.set(Math.floor(step / Math.pow(2, i)) % 2 ? 1.0 : -1.0, i);
-    }
-    return buffer.toTensor().as1D();
+    return buffer.toTensor().as2D(numSteps, this.depth);
```

Comment:
Should it be an array of 1D? Is this just going to have to be split later?

Reply:
Talked offline, will revisit later.